### DISCUSSION TITLE
MyApp.getInitialProps + TRPC

### DISCUSSION THREAD

**Discussion by jacksonblankenship** [trpc/trpc#1710] [Q&A] [answered]
My initial version of this question was way too verbose. I've simplified the problem I'm facing [below](https://github.com/trpc/trpc/discussions/1710#discussioncomment-2598155).

**Comment by sachinraja** [comment]
tRPC does set a `getInitialProps` but this *should* work. It may be a bug, not sure. Got a reproduction?

**Comment by KATT** [comment]
I think this should do the job for you: 

https://github.com/trpc/trpc/blob/f94a0699eb726b68e46e27ed4b48a26b5277d06b/examples/next-prisma-starter-websockets/src/pages/_app.tsx#L74-L83

In the [WebSockets example](https://github.com/trpc/examples-next-prisma-starter-websockets) we do authed SSR with cookies

**Comment by base21** [comment]
I get this same error as soon as I upgrade to react-18 
![image](https://user-images.githubusercontent.com/84002624/161602705-62cbd78f-b95c-4d50-8e02-5784c5891e9a.png)

**Comment by jacksonblankenship** [comment]
Hey @KATT, thanks for getting back. Unfortunately, this solution didn't seem to solve my problem. I also feel as though I overcomplicated my question by including the mantine-related stuff. I'm sorry about that.

Here's the core of my issue. I need to be able to pass `pageProps` to my custom `_app.tsx` but for some reason, trpc is blocking me from doing this. I'm still assuming this is because of the `withTRPC()` wrapper in `_app.tsx`, which is [noted in the docs](https://trpc.io/docs/ssr#caveats), but I'm curious if there is a way around this.

Per @sachinraja's suggestion and to clear things up a bit, I'm including a code sandbox link that just works off the basic trpc example and attempts to do `getInitialProps`. You'll see me attempting to pass `someInitialProp` to `MyApp`'s pageProps via `MyApp.getInitialProps`, but the value doesn't ever arrive.

https://codesandbox.io/s/trpc-custom-app-getinitialprops-yo8pd6?file=/src/pages/_app.tsx

## Found
```
{
  trpcState:  {
    json: { ... },
    meta: { ... },
  },
}
```

## Expected
```
{
  trpcState:  {
    json: { ... },
    meta: { ... },
  },
  someInitialProp: "some value",
}
```

Hopefully, this will help us get closer to an answer.

**Reply by KATT** [reply]
Seems related to #596

**Comment by needim** [comment, marked as answer]
HOC captures pageProps if it is an App component.

https://github.com/trpc/trpc/blob/98568b5914df0f97936d6ec0cefde425dba9ba6a/packages/next/src/withTRPC.tsx#L158-L170

So you should return pageProps from your getInitialProps function

example:
```typescript

const MyApp = (
  props: AppProps & { session: Session | null; colorScheme: ColorScheme }
) => {
  const { Component, pageProps } = props;
  const [colorScheme, setColorScheme] = useState<ColorScheme>(
    pageProps.colorScheme
  );

...
...
...

MyApp.getInitialProps = async (appContext: AppContext) => {
  const color = getCookie("mantine-color-scheme", appContext.ctx);

  return { pageProps: { colorScheme: color || "dark" } };
};
```

**Reply by mickasmt** [reply]
@needim I was looking for this solution to save the color scheme in a cookie with Mantine.
This solution works perfectly for me!
Thank you very much!

**Reply by nathanqueija** [reply]
@needim my hero! Thanks

**Reply by larisonjustin** [reply]
I spent hours trying to solve this issue, now I can go to sleep. @needim, thank you my friend!
